File: python/benchmark_test.py
import sys
from random import sample
import timeit
from util import Point
from quadratic_sol import quadratic_solution
from nlogn_sol import nlogn_solution
from multiproc_sol import nlogn_solution_multiproc
from file_test import read_test_file


def random_sample(size):
    x = sample(range(size*100), size)
    y = sample(range(size*100), size)
    return [Point(x, y) for x, y in list(zip(x, y))]


if __name__ == "__main__":

    def main():
        if len(sys.argv) < 2:
            print("Usage: python benchmark_test.py test_mode [sample_size|filename]")
            exit(1)

        test_mode = sys.argv[1]
        if test_mode == 'random':
            if len(sys.argv) < 3:
                print("Usage: python benchmark_test.py 'random' sample_size")
                exit(1)
            P = random_sample(int(sys.argv[2]))
        elif test_mode == 'file':
            if len(sys.argv) < 3:
                print("Usage: python benchmark_test.py 'file' filename")
                exit(1)
            P = read_test_file(sys.argv[2])
        else:
            print("test mode must be one of: 'random', 'file'")
            exit(1)

        print(f"nlogn_solution {timeit.timeit(lambda: nlogn_solution(P), number=1)}")
        print(f"nlogn_solution_par 4 {timeit.timeit(lambda: nlogn_solution_multiproc(P, 4), number=1)}")
        print(f"nlogn_solution_par 8 {timeit.timeit(lambda: nlogn_solution_multiproc(P, 8), number=1)}")

    main()

File: c/benchmark_test.c
#include "closest_point.h"

//gcc benchmark_test.c util.c nlogn_sol.c quadratic_sol.c multiproc_sol.c multithread_sol.c -o out/benchmark_test
int main(int argc, char const *argv[])
{
    int num_points;
    struct pointlist pd;

    if (argc < 2)
    {
        printf("USAGE: %s test_mode [sample_size|filename]", argv[0]);
        return EXIT_FAILURE;
    }

    const char* test_mode = argv[1];
    if(strcmp(test_mode, "random") == 0) {
        if(argc < 3) {
            printf("Usage: %s 'random' sample_size", argv[0]);
            return EXIT_FAILURE;
        }
        num_points = atoi(argv[2]);
        point* P = rand_points(num_points);
        pd.num_points = num_points;
        pd.P = P;
    }
    else if(strcmp(test_mode, "file") == 0) {
        if(argc < 3) {
            printf("Usage: %s 'file' filename", argv[0]);
            return EXIT_FAILURE;
        }
        pd = read_test_file(argv[2]);
    }
    else{
        printf("test mode must be one of: 'random', 'file'");
        return EXIT_FAILURE;
    }

    printf("nlogn_solution: num_points=%ld, num_processes=%d, time=%.4f seconds\n", pd.num_points, -1, timeit(nlogn_solution, pd, -1));
    printf("nlogn_solution_multiproc: num_points=%ld, num_processes=%d, time=%.4f seconds\n", pd.num_points, 2, timeit(nlogn_solution_multiproc, pd, 8));
    printf("nlogn_solution_multithread: num_points=%ld, num_processes=%d, time=%.4f seconds\n", pd.num_points, 2, timeit(nlogn_solution_multithread, pd, 8));
    printf("quadratic_solution: num_points=%ld, num_processes=%d, time=%.4f seconds\n", pd.num_points, -1, timeit(quadratic_solution, pd, -1));

    return EXIT_SUCCESS;
}
